File: github-pr-merger.js
#!/usr/bin/env node

const https = require('https');
const { execSync } = require('child_process');

console.log('🚀 GitHub PR Merger - Starting...');

// Configuration
const REPO_OWNER = 'Zion-Holdings';
const REPO_NAME = 'zion.app';
const GITHUB_TOKEN = process.env.GITHUB_TOKEN || '';

// Function to make GitHub API requests
function githubRequest(path, method = 'GET', data = null) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'api.github.com',
      port: 443,
      path: path,
      method: method,
      headers: {
        'User-Agent': 'PR-Merger-Script',
        'Accept': 'application/vnd.github.v3+json',
        'Authorization': GITHUB_TOKEN ? `token ${GITHUB_TOKEN}` : '',
      }
    };

    if (data) {
      options.headers['Content-Type'] = 'application/json';
      options.headers['Content-Length'] = Buffer.byteLength(data);
    }

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        try {
          const jsonBody = JSON.parse(body);
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(jsonBody);
          } else {
            reject(new Error(`GitHub API error: ${res.statusCode} - ${jsonBody.message || body}`));
          }
        } catch (e) {
          reject(new Error(`Failed to parse response: ${body}`));
        }
      });
    });

    req.on('error', reject);
    
    if (data) {
      req.write(data);
    }
    req.end();
  });
}

// Function to run git commands
function runGit(command) {
  try {
    return execSync(command, { 
      cwd: '/workspace',
      encoding: 'utf8',
      stdio: 'pipe'
    }).trim();
  } catch (error) {
    throw new Error(`Git command failed: ${command} - ${error.message}`);
  }
}

// Function to get open PRs
async function getOpenPRs() {
  console.log('📋 Fetching open PRs...');
  try {
    const prs = await githubRequest(`/repos/${REPO_OWNER}/${REPO_NAME}/pulls?state=open&per_page=100`);
    console.log(`Found ${prs.length} open PRs`);
    return prs;
  } catch (error) {
    console.error('❌ Failed to fetch PRs:', error.message);
    return [];
  }
}

// Function to check if PR can be merged
async function checkPRMergeable(prNumber) {
  try {
    const pr = await githubRequest(`/repos/${REPO_OWNER}/${REPO_NAME}/pulls/${prNumber}`);
    return pr.mergeable;
  } catch (error) {
    console.error(`❌ Failed to check mergeable status for PR #${prNumber}:`, error.message);
    return false;
  }
}

// Function to merge a PR
async function mergePR(prNumber, title) {
  try {
    console.log(`🔄 Merging PR #${prNumber}: ${title}`);
    
    const mergeData = JSON.stringify({
      commit_title: `Merge PR #${prNumber}: ${title}`,
      commit_message: `Automated merge of PR #${prNumber}`,
      merge_method: 'merge'
    });
    
    const result = await githubRequest(
      `/repos/${REPO_OWNER}/${REPO_NAME}/pulls/${prNumber}/merge`,
      'PUT',
      mergeData
    );
    
    console.log(`✅ Successfully merged PR #${prNumber}`);
    return true;
  } catch (error) {
    console.error(`❌ Failed to merge PR #${prNumber}:`, error.message);
    return false;
  }
}

// Function to resolve conflicts locally
async function resolveConflictsLocally(pr) {
  try {
    console.log(`🔧 Attempting to resolve conflicts for PR #${pr.number} locally...`);
    
    // Switch to main
    runGit('git checkout main');
    runGit('git pull origin main');
    
    // Fetch the PR branch
    runGit(`git fetch origin ${pr.head.ref}`);
    
    // Try to merge
    try {
      runGit(`git merge origin/${pr.head.ref} --no-ff -m "Merge PR #${pr.number}: ${pr.title}"`);
      console.log(`✅ Successfully merged PR #${pr.number} locally`);
      
      // Push to main
      runGit('git push origin main');
      return true;
    } catch (mergeError) {
      console.log(`⚠️  Merge conflicts detected for PR #${pr.number}`);
      
      // Get conflicted files
      const conflictFiles = runGit('git diff --name-only --diff-filter=U').split('\n').filter(f => f);
      
      if (conflictFiles.length > 0) {
        console.log(`🔧 Resolving conflicts in ${conflictFiles.length} files...`);
        
        // Resolve conflicts by keeping our changes
        for (const file of conflictFiles) {
          try {
            runGit(`git checkout --ours "${file}"`);
            runGit(`git add "${file}"`);
            console.log(`✅ Resolved conflicts in: ${file}`);
          } catch (fileError) {
            console.log(`⚠️  Could not resolve conflicts in: ${file}`);
          }
        }
        
        // Commit the resolved conflicts
        runGit(`git commit -m "Resolve merge conflicts for PR #${pr.number}"`);
        runGit('git push origin main');
        
        console.log(`✅ Successfully resolved conflicts and merged PR #${pr.number}`);
        return true;
      } else {
        console.log(`❌ No conflict files found for PR #${pr.number}`);
        runGit('git merge --abort');
        return false;
      }
    }
  } catch (error) {
    console.error(`❌ Failed to resolve conflicts for PR #${pr.number}:`, error.message);
    return false;
  }
}

// Main function
async function main() {
  try {
    console.log('📍 Repository:', `${REPO_OWNER}/${REPO_NAME}`);
    
    // Get open PRs
    const openPRs = await getOpenPRs();
    
    if (openPRs.length === 0) {
      console.log('✅ No open PRs found');
      return;
    }
    
    let mergedCount = 0;
    let conflictCount = 0;
    
    // Process each PR
    for (const pr of openPRs) {
      console.log(`\n🔄 Processing PR #${pr.number}: ${pr.title}`);
      
      // Check if PR can be merged
      const mergeable = await checkPRMergeable(pr.number);
      
      if (mergeable === true) {
        // PR can be merged directly
        if (await mergePR(pr.number, pr.title)) {
          mergedCount++;
        }
      } else if (mergeable === false) {
        // PR has conflicts, try to resolve locally
        conflictCount++;
        if (await resolveConflictsLocally(pr)) {
          mergedCount++;
        }
      } else {
        console.log(`⚠️  PR #${pr.number} mergeable status is unknown, skipping...`);
      }
      
      // Small delay between PRs
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
    
    console.log(`\n🎉 Process completed!`);
    console.log(`✅ Successfully merged: ${mergedCount} PRs`);
    console.log(`⚠️  Conflicts encountered: ${conflictCount} PRs`);
    
  } catch (error) {
    console.error('❌ Error during PR merge process:', error.message);
    process.exit(1);
  }
}

// Run the main function
main();

File: github_pr_merger.py
#!/usr/bin/env python3

"""
GitHub PR Merger Script
This script will find and merge all open PRs into the main branch
"""

import subprocess
import sys
import os
import json
import re
from datetime import datetime
from pathlib import Path

class GitHubPRMerger:
    def __init__(self):
        self.main_branch = "main"
        self.remote = "origin"
        self.log_file = f"pr_merge_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
        
    def log(self, message):
        """Log with timestamp"""
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        log_message = f"[{timestamp}] {message}"
        print(log_message)
        
        # Write to log file
        with open(self.log_file, 'a') as f:
            f.write(log_message + '\n')
    
    def run_command(self, cmd, check=True):
        """Run a command and return the result"""
        try:
            self.log(f"Executing: {cmd}")
            result = subprocess.run(cmd, shell=True, capture_output=True, text=True, check=check)
            return result.stdout.strip(), result.stderr.strip()
        except subprocess.CalledProcessError as e:
            self.log(f"Command failed: {cmd}")
            self.log(f"Error: {e.stderr}")
            return e.stdout.strip(), e.stderr.strip()
    
    def check_git_repo(self):
        """Check if we're in a git repository"""
        if not os.path.exists(".git"):
            self.log("❌ Not in a git repository")
            sys.exit(1)
        self.log("✅ Git repository found")
    
    def get_current_branch(self):
        """Get current branch name"""
        stdout, stderr = self.run_command("git branch --show-current")
        self.log(f"📍 Current branch: {stdout}")
        return stdout
    
    def check_merge_conflicts(self):
        """Check for merge conflicts"""
        self.log("🔍 Checking for merge conflicts...")
        stdout, stderr = self.run_command("git diff --name-only --diff-filter=U")
        
        if stdout:
            self.log("⚠️  Found merge conflicts in:")
            for file in stdout.split('\n'):
                if file.strip():
                    self.log(f"  - {file}")
            return stdout.split('\n')
        else:
            self.log("✅ No merge conflicts found")
            return []
    
    def resolve_merge_conflicts(self, conflicted_files):
        """Resolve merge conflicts automatically"""
        self.log("🔧 Resolving merge conflicts...")
        
        for file in conflicted_files:
            if not file.strip():
                continue
                
            self.log(f"📝 Resolving: {file}")
            
            # Create backup
            backup_name = f"{file}.backup.{int(datetime.now().timestamp())}"
            self.run_command(f"cp '{file}' '{backup_name}'")
            
            # Use git checkout to take current branch version
            self.run_command(f"git checkout --ours '{file}'")
            
            # Add resolved file
            self.run_command(f"git add '{file}'")
        
        self.log("✅ Merge conflicts resolved")
    
    def get_all_branches(self):
        """Get all remote branches"""
        self.log("📋 Getting all remote branches...")
        stdout, stderr = self.run_command("git branch -r --sort=-committerdate")
        
        branches = []
        for line in stdout.split('\n'):
            if line.strip() and 'origin/main' not in line:
                clean_branch = line.strip().replace('origin/', '')
                branches.append(clean_branch)
        
        self.log(f"Found {len(branches)} branches to process")
        return branches[:20]  # Limit to 20 most recent
    
    def merge_branch(self, branch):
        """Merge a specific branch"""
        self.log(f"🔄 Attempting to merge branch: {branch}")
        
        # Fetch the branch
        self.run_command(f"git fetch {self.remote} {branch}")
        
        # Try to merge
        stdout, stderr = self.run_command(f"git merge {self.remote}/{branch} --no-edit", check=False)
        
        if "CONFLICT" in stderr or "conflict" in stderr:
            self.log(f"⚠️  Merge conflicts in {branch}, resolving...")
            
            # Get conflicted files
            conflicted_files = self.check_merge_conflicts()
            
            if conflicted_files:
                self.resolve_merge_conflicts(conflicted_files)
                
                # Commit resolved conflicts
                stdout, stderr = self.run_command("git commit -m 'Resolve merge conflicts automatically'", check=False)
                self.log(f"✅ Resolved conflicts for {branch}")
            
            return True
        else:
            self.log(f"✅ Successfully merged {branch}")
            return True
    
    def merge_with_main(self):
        """Merge with main branch"""
        self.log("🔄 Attempting to merge with main...")
        
        # Fetch latest changes
        self.run_command(f"git fetch {self.remote} {self.main_branch}")
        
        # Try to merge
        stdout, stderr = self.run_command(f"git merge {self.remote}/{self.main_branch} --no-edit", check=False)
        
        if "CONFLICT" in stderr or "conflict" in stderr:
            self.log("⚠️  Merge conflicts detected")
            return False
        else:
            self.log("✅ Successfully merged with main")
            return True
    
    def push_changes(self, branch):
        """Push changes to remote"""
        self.log(f"📤 Pushing changes for branch: {branch}")
        
        stdout, stderr = self.run_command(f"git push {self.remote} {branch}", check=False)
        
        if "error" in stderr.lower() or "failed" in stderr.lower():
            self.log(f"❌ Error pushing {branch}: {stderr}")
            return False
        else:
            self.log(f"✅ Successfully pushed {branch}")
            return True
    
    def process_all_branches(self):
        """Process all branches for merging"""
        self.log("🔄 Processing all branches for merging...")
        
        # Get all branches
        branches = self.get_all_branches()
        
        # Switch to main first
        self.run_command(f"git checkout {self.main_branch}")
        self.run_command(f"git pull {self.remote} {self.main_branch}")
        
        # Process each branch
        for branch in branches:
            if branch != self.main_branch:
                self.log(f"🔄 Processing branch: {branch}")
                
                if self.merge_branch(branch):
                    self.log(f"✅ Successfully processed {branch}")
                else:
                    self.log(f"❌ Failed to process {branch}")
        
        # Push main branch
        self.push_changes(self.main_branch)
    
    def cleanup_merged_branches(self):
        """Clean up merged branches"""
        self.log("🧹 Cleaning up merged branches...")
        
        # Switch to main
        self.run_command(f"git checkout {self.main_branch}")
        self.run_command(f"git pull {self.remote} {self.main_branch}")
        
        # Delete merged branches
        stdout, stderr = self.run_command("git branch --merged")
        
        for line in stdout.split('\n'):
            if line.strip() and 'main' not in line and '*' not in line:
                branch = line.strip()
                self.run_command(f"git branch -d {branch}", check=False)
                self.log(f"🗑️  Deleted merged branch: {branch}")
        
        self.log("✅ Cleanup completed")
    
    def run(self):
        """Main execution function"""
        self.log("🚀 Starting GitHub PR Merger...")
        
        # Check if we're in a git repo
        self.check_git_repo()
        
        # Get current branch
        current_branch = self.get_current_branch()
        
        # Check for existing merge conflicts
        conflicted_files = self.check_merge_conflicts()
        if conflicted_files:
            self.resolve_merge_conflicts(conflicted_files)
            self.run_command("git commit -m 'Resolve existing merge conflicts automatically'", check=False)
        
        # Process all branches
        self.process_all_branches()
        
        # Clean up
        self.cleanup_merged_branches()
        
        self.log("🎉 GitHub PR Merger completed successfully!")
        self.log(f"📋 Log saved to: {self.log_file}")

if __name__ == "__main__":
    merger = GitHubPRMerger()
    merger.run()